lib/TH/THGeneral.c

```diff
@@ -137,10 +159,17 @@ void* THRealloc(void *ptr, long size)
   if(size < 0)
     THError("$ Torch: invalid memory size -- maybe an overflow?");
 
-  ptr = realloc(ptr, size);
-  if(!ptr)
+  void *newptr = realloc(ptr, size);
+
+  if(!newptr && torchGCFunction) {
+    torchGCFunction(torchGCData);
+    newptr = realloc(ptr, size);
+  }
+
+  if(!newptr)
     THError("$ Torch: not enough memory: you tried to reallocate %dGB. Buy new RAM!", size/1073741824);
```

Comment:
@dominikgrewe I didn't say the process would exit, just that the function would never return. For example, if the caller writes `self->data = THRealloc(self->data, size)`, and you hit the error, `self->data` will not be modified and will be freed when `self` is cleaned up.
Notice that I haven't changed the behavior of THRealloc w.r.t. OOM - it always just called THError.

@tkoeppe no, `THRealloc` does not have the same semantics as `realloc`. The semantics are "if `THRealloc` ever returns, then it returns a pointer to a reallocated block of the requested size".